how-it-works/1_protocol/04_execution.card.md

```diff
@@ -6,10 +6,20 @@ title: Execution Layer
 
 # Execution layer
```

Comment:
@dietersommer: I didn't see that you've rewritten this part a bit and did a rewrite of my own. It has some of the ideas held here previously and a slightly different organization. 

The execution layer is the topmost layer of the core IC protocol stack. This layer implements a Web Assembly (Wasm) virtual machine used to execute canister code – recall that canisters are essentially Wasm programs.  

Canister execution comes in two forms, each of which represents a trade-off between performance and security. Non-replicated execution is similar in speed to Web access. This type of execution is fast since it takes place on a single node of a subnetwork, so by default the results obtained this way are not trustworthy. In contrast, replicated execution takes place on all nodes of the subnetwork.  Here, the execution layer works closely with message routing to ensure that nodes perform precisely the same computation and therefore that they make exactly the same changes on the state of the canisters that they run.  In other words, message routing and execution implement a replicated state machine, a key property of a blockchain. 

In a bit more detail, replicated computation proceeds in rounds. In every round, the message routing layer gets a block of messages from consensus, it puts these messages into the input queues of their destination canisters and then hands over control to the execution layer.  
This layer determines an order in which to execute the canisters and then schedules either until all messages in the canisters' queues are consumed or the instruction limit for the round has been reached, to ensure bounded round times. 

The execution layer implements three features which set apart the IC from other blockchains. 
First, canisters can be executed concurrently.  This is possible since canisters depend on one another only through the messages that they exchange and this communication is asynchronous.  
Second, canister execution can span multiple rounds through a novel technique called *deterministic time slicing (DTS*). That is, execution of canisters which requires more instructions than available in a round is paused and continues in the subsequent round.  
 Finally, although replicated computation needs deterministic canisters can use an pseudorandom number generator completely unpredictable from one round to the next.